File: rider/src/main/java/com/pss9/rideraid/ant/AntGarminGSC10.java
package com.pss9.rideraid.ant;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import com.dsi.ant.plugins.antplus.pcc.AntPlusBikeCadencePcc;
import com.dsi.ant.plugins.antplus.pcc.AntPlusBikeSpeedDistancePcc;
import com.dsi.ant.plugins.antplus.pcc.defines.DeviceState;
import com.dsi.ant.plugins.antplus.pcc.defines.EventFlag;
import com.dsi.ant.plugins.antplus.pcc.defines.RequestAccessResult;
import com.dsi.ant.plugins.antplus.pccbase.AntPluginPcc;
import com.dsi.ant.plugins.antplus.pccbase.AntPlusBikeSpdCadCommonPcc;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.EnumSet;

public class AntGarminGSC10 extends AntBikeDevice {
    protected ArrayList<AntPlusBikeSpdCadCommonPcc> antPCC = new ArrayList<>();

    private boolean active;

    private AntPlusBikeCadencePcc bcPcc = null;
    private AntPlusBikeSpeedDistancePcc bsPcc = null;

    public AntGarminGSC10(@Tire.TireCircumferenceMM long tire) {
        super(tire);

        antDeviceType = ANT_DEVICE_TYPE_CADENCE;
    }

    @Override
    public void release() {
        for (AntPlusBikeSpdCadCommonPcc pcc : antPCC) {
            pcc.releaseAccess();
        }

        antPCC.clear();
        antPCC = null;

        active = false;
    }

    @Override
    public boolean isActive() {
        return active;
    }

    private String processReturnCode(int returnCode) {
        switch (RequestAccessResult.getValueFromInt(returnCode)) {
            case CHANNEL_NOT_AVAILABLE:
                return "Channel Not Available";
            case OTHER_FAILURE:
                return "RequestAccess failed";
            case USER_CANCELLED:
                return "Cancelled by user";
            case UNRECOGNIZED:
                return "Failed: UNRECOGNIZED. Upgrade Required?";
        }

        return "Unknown result: " + returnCode;
    }

    public void activate(final Activity activity) {
        AntPlusBikeCadencePcc.requestAccess(
                activity,
                activity.getApplicationContext(),

                new AntPluginPcc.IPluginAccessResultReceiver<AntPlusBikeCadencePcc>() {
                    //Handle the result, connecting to events on success or reporting failure to user.
                    @Override
                    public void onResultReceived(AntPlusBikeCadencePcc result,
                                                 RequestAccessResult resultCode,
                                                 DeviceState initialDeviceState) {
                        String status = "";
                        switch (resultCode) {
                            case SUCCESS:
                                antPCC.add(result);

                                subscribeCadenceEvent(result);

                                activateCombinedSpeedSensor(activity.getApplicationContext(), result);

                                active = true;
                                eventHandler.onEvent(createEvent(ANT_DEVICE_ACTIVE, 0));
                                break;
                            case DEPENDENCY_NOT_INSTALLED:
                                AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(activity);
                                alertDialogBuilder.setTitle("Missing Service");
                                alertDialogBuilder.setMessage("\"" + AntPlusBikeCadencePcc.getMissingDependencyName() +
                                        "\" was not found.\nPlease install the latest ANT+ Plugins service. " +
                                        "Would you like to launch the Play Store now?");
                                alertDialogBuilder.setCancelable(true);
                                alertDialogBuilder.setPositiveButton("Go to Store", new DialogInterface.OnClickListener() {
                                    @Override
                                    public void onClick(DialogInterface dialog, int which) {
                                        Intent startStore = new Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=" +
                                                AntPlusBikeCadencePcc.getMissingDependencyPackageName()));
                                        startStore.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);

                                        activity.startActivity(startStore);
                                        dialog.dismiss();
                                    }
                                });
                                alertDialogBuilder.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
                                    @Override
                                    public void onClick(DialogInterface dialog, int which) {
                                        dialog.dismiss();
                                    }
                                });

                                alertDialogBuilder.create().show();
                                break;
                            default:
                                eventHandler.onEvent(createEvent(ANT_DEVICE_ERROR, resultCode.getIntValue()));
                                break;
                        }
                    }
                },

                newDeviceState -> {
                    if (newDeviceState == DeviceState.DEAD) {
                        eventHandler.onEvent(createEvent(ANT_DEVICE_ERROR, DeviceState.DEAD.getIntValue()));
                    }
                });
    }

    private void subscribeCadenceEvent(final AntPlusBikeSpdCadCommonPcc pcc) {
        ((AntPlusBikeCadencePcc) pcc).subscribeCalculatedCadenceEvent(
                new AntPlusBikeCadencePcc.ICalculatedCadenceReceiver() {
                    @Override
                    public void onNewCalculatedCadence(final long estTimestamp,
                                                       final EnumSet<EventFlag> eventFlags,
                                                       final BigDecimal calculatedCadence) {
                        eventHandler.onEvent(createEvent(ANT_DEVICE_TYPE_CADENCE, calculatedCadence.longValue()));
                    }
                });
    }

    private void activateCombinedSpeedSensor(Context context, final AntPlusBikeSpdCadCommonPcc pcc) {
        if (!pcc.isSpeedAndCadenceCombinedSensor()) {
            return;
        }

        AntPlusBikeSpeedDistancePcc.requestAccess(context, pcc.getAntDeviceNumber(), 0, true,
                new AntPluginPcc.IPluginAccessResultReceiver<AntPlusBikeSpeedDistancePcc>() {
                    //Handle the result, connecting to events on success or reporting failure to user.
                    @Override
                    public void onResultReceived(AntPlusBikeSpeedDistancePcc result,
                                                 RequestAccessResult resultCode,
                                                 DeviceState initialDeviceStateCode) {
                        switch (resultCode) {
                            case SUCCESS:
                                antPCC.add(result);
                                result.subscribeCalculatedSpeedEvent(
                                        new AntPlusBikeSpeedDistancePcc.CalculatedSpeedReceiver(bikeTireCircumference) {
                                            @Override
                                            public void onNewCalculatedSpeed(long estTimestamp,
                                                                             EnumSet<EventFlag> eventFlags,
                                                                             final BigDecimal calculatedSpeed) {
                                                // units m/s
                                                eventHandler.onEvent(createEvent(ANT_DEVICE_TYPE_SPEED, calculatedSpeed.longValue()));
                                            }
                                        });

                                eventHandler.onEvent(createEvent(ANT_DEVICE_ACTIVE, 0));
                                break;
                            default:
                                eventHandler.onEvent(createEvent(ANT_DEVICE_ERROR, resultCode.getIntValue()));
                                break;
                        }
                    }
                },

                //Receives state changes and shows it on the status display line
                new AntPluginPcc.IDeviceStateChangeReceiver() {
                    @Override
                    public void onDeviceStateChange(final DeviceState newDeviceState) {
                        if (newDeviceState == DeviceState.DEAD) {
                            eventHandler.onEvent(createEvent(ANT_DEVICE_ERROR, DeviceState.DEAD.getIntValue()));
                        }
                    }
                }
        );
    }
}

File: rider/src/main/java/com/pss9/rider/ant/AntGarminGSC10.kt
package com.pss9.rider.ant

import android.app.Activity
import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.net.Uri
import com.dsi.ant.plugins.antplus.pcc.AntPlusBikeCadencePcc
import com.dsi.ant.plugins.antplus.pcc.AntPlusBikeSpeedDistancePcc
import com.dsi.ant.plugins.antplus.pcc.defines.DeviceState
import com.dsi.ant.plugins.antplus.pcc.defines.EventFlag
import com.dsi.ant.plugins.antplus.pcc.defines.RequestAccessResult
import com.dsi.ant.plugins.antplus.pccbase.AntPlusBikeSpdCadCommonPcc
import java.math.BigDecimal
import java.util.*

class AntGarminGSC10(tire: Long) : AntBikeDevice(tire) {
    private var antPCC: ArrayList<AntPlusBikeSpdCadCommonPcc>? = ArrayList()

    init {
        type = AntDevice.ANT_DEVICE_TYPE_CADENCE
    }

    override fun activate(activity: Activity) {
        AntPlusBikeCadencePcc.requestAccess(
            activity,
            activity.applicationContext,
            { result, resultCode, _ ->
                //Handle the result, connecting to events on success or reporting failure to user.
                when (resultCode) {
                    RequestAccessResult.SUCCESS -> {
                        antPCC!!.add(result)

                        subscribeCadenceEvent(result)

                        activateCombinedSpeedSensor(activity.applicationContext, result)

                        isActive = true
                        eventHandler.onEvent(BikeEvent.createEvent(AntDevice.ANT_DEVICE_ACTIVE, 0))
                    }
                    RequestAccessResult.DEPENDENCY_NOT_INSTALLED -> {
                        val alertDialogBuilder = AlertDialog.Builder(activity)
                        alertDialogBuilder.setTitle("Missing Service")
                        alertDialogBuilder.setMessage(
                            "\"" + AntPlusBikeCadencePcc.getMissingDependencyName() +
                                    "\" was not found.\nPlease install the latest ANT+ Plugins service. " +
                                    "Would you like to launch the Play Store now?"
                        )
                        alertDialogBuilder.setCancelable(true)
                        alertDialogBuilder.setPositiveButton("Go to Store") { dialog, _ ->
                            val startStore = Intent(
                                Intent.ACTION_VIEW,
                                Uri.parse("market://details?id=" + AntPlusBikeCadencePcc.getMissingDependencyPackageName())
                            )
                            startStore.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)

                            activity.startActivity(startStore)
                            dialog.dismiss()
                        }
                        alertDialogBuilder.setNegativeButton("Cancel") { dialog, _ -> dialog.dismiss() }

                        alertDialogBuilder.create().show()
                    }
                    else -> eventHandler.onEvent(
                        BikeEvent.createEvent(
                            AntDevice.ANT_DEVICE_ERROR,
                            resultCode.intValue.toLong()
                        )
                    )
                }
            }, { state ->
                if (state == DeviceState.DEAD) {
                    eventHandler.onEvent(
                        BikeEvent.createEvent(
                            AntDevice.ANT_DEVICE_ERROR,
                            DeviceState.DEAD.intValue.toLong()
                        )
                    )
                }
            })
    }

    override fun release() {
        for (pcc in antPCC!!) {
            pcc.releaseAccess()
        }

        antPCC!!.clear()
        antPCC = null

        isActive = false
    }

    private fun subscribeCadenceEvent(pcc: AntPlusBikeSpdCadCommonPcc) {
        (pcc as AntPlusBikeCadencePcc).subscribeCalculatedCadenceEvent { _, _, calculatedCadence ->
            eventHandler.onEvent(
                BikeEvent.createEvent(AntDevice.ANT_DEVICE_TYPE_CADENCE, calculatedCadence.toLong())
            )
        }
    }

    private fun activateCombinedSpeedSensor(context: Context, pcc: AntPlusBikeSpdCadCommonPcc) {
        if (!pcc.isSpeedAndCadenceCombinedSensor) {
            return
        }

        AntPlusBikeSpeedDistancePcc.requestAccess(context, pcc.antDeviceNumber, 0, true,
            { result, resultCode, _ ->
                //Handle the result, connecting to events on success or reporting failure to user.
                when (resultCode) {
                    RequestAccessResult.SUCCESS -> {
                        antPCC!!.add(result)
                        result.subscribeCalculatedSpeedEvent(
                            object : AntPlusBikeSpeedDistancePcc.CalculatedSpeedReceiver(bikeTireCircumference) {
                                override fun onNewCalculatedSpeed(
                                    estTimestamp: Long,
                                    eventFlags: EnumSet<EventFlag>,
                                    calculatedSpeed: BigDecimal
                                ) {
                                    // units m/s
                                    eventHandler.onEvent(
                                        BikeEvent.createEvent(
                                            AntDevice.ANT_DEVICE_TYPE_SPEED,
                                            calculatedSpeed.toLong()
                                        )
                                    )
                                }
                            })

                        eventHandler.onEvent(BikeEvent.createEvent(AntDevice.ANT_DEVICE_ACTIVE, 0))
                    }
                    else -> eventHandler.onEvent(
                        BikeEvent.createEvent(
                            AntDevice.ANT_DEVICE_ERROR,
                            resultCode.intValue.toLong()
                        )
                    )
                }
            },

            //Receives state changes and shows it on the status display line
            { newDeviceState ->
                if (newDeviceState == DeviceState.DEAD) {
                    eventHandler.onEvent(
                        BikeEvent.createEvent(
                            AntDevice.ANT_DEVICE_ERROR,
                            DeviceState.DEAD.intValue.toLong()
                        )
                    )
                }
            }
        )
    }

    private fun processReturnCode(returnCode: Int): String =
        when (RequestAccessResult.getValueFromInt(returnCode)) {
            RequestAccessResult.CHANNEL_NOT_AVAILABLE -> "Channel Not Available"
            RequestAccessResult.OTHER_FAILURE -> "RequestAccess failed"
            RequestAccessResult.USER_CANCELLED -> "Cancelled by user"
            RequestAccessResult.UNRECOGNIZED -> "Failed: UNRECOGNIZED. Upgrade Required?"
            else -> "Unknown result: $returnCode"
        }

}
